Feature/migration

suisseWalter opened the pull request:
automatic save file migration option.

suisseWalter committed:
added factorio_version function
Before the solution involved running a full benchmark to extract the version number.
Also fixed flake8 errors introduced in 4312f97.

suisseWalter committed:
Update benchmarker.py

fixed improper build detection

Co-authored-by: flameSla <39356103+flameSla@users.noreply.github.com>

suisseWalter committed:
first part of migration
still highly broken, causes random factorio process to fly around.
and other weird stuff. but overall it does what it's supposed to do.

suisseWalter committed:
map migration
Using psutil there are no problems with terminating factorio.

suisseWalter committed:
Merge branch 'master' into feature/migration

suisseWalter committed:
updated readme, added custom-script
enables the running of custom scripts upon migration.

suisseWalter committed:
Merge branch 'feature/migration' of github.com:suisseWalter/factorio-benchmark into feature/migration

flameSla commented on benchmarker.py:
```diff
@@ -554,16 +654,32 @@ def init_parser() -> argparse.ArgumentParser:
+        exit()
 
-    create_mods_dir()
     if args.disable_mods:
```
imho it is better to leave the call to the **create_mods_dir()** function
it will be required on Windows
and on Linux after manually installing Factorio

flameSla commented:
factorio_rcon -> requirements.txt

flameSla commented on benchmarker.py:
```diff
@@ -71,6 +84,61 @@ def get_factorio_version(factorio_bin: str, full: bool = False) -> str:
+    if not inplace:
+        oldmap = Path(map)
+        newmap = Path(
+            os.path.splitext(map)[0] + get_factorio_version(factorio_bin).replace(".", "") + ".zip"
```
```suggestion
            os.path.splitext(map)[0] + "_" + get_factorio_version(factorio_bin).replace(".", "_") + ".zip"
```
imho so the version will be clearer

suisseWalter replied:
but longer. yes it's more readable, but the filenames are already really long most of the time. and so adding even more (especially with multiple migrations) isn't optimal.

suisseWalter committed:
updated requirements.txt

flameSla commented:
how do I migrate from version 0.16?

suisseWalter commented:
replace the link with the proper version for windows. (probably full/win)
also you might have to remove the factorio folder before running `-u`. 
```
benchmarker -u --version_link https://www.factorio.com/get-download/1.0.0/headless/linux64 
benchmarker -mi
benchmarker -u
bemchmarker -mi 
```

rpdelaney reviewed (changes requested):
We should never have `os.path` operations inside a `pathlib.Path`.. the latter can do everything with its built-in methods.

rpdelaney commented on benchmarker.py:
```diff
@@ -71,6 +84,61 @@ def get_factorio_version(factorio_bin: str, full: bool = False) -> str:
 
+def migrate_map(factorio_bin: str, map: str, inplace: bool, custom_script: str | None) -> None:
+    """migrate map to a new factorio version."""
+    print("migrating map:" + map)
```
```suggestion
    print(f"migrating map: {map}")
```
nit

rpdelaney commented on benchmarker.py:
```diff
@@ -71,6 +84,61 @@ def get_factorio_version(factorio_bin: str, full: bool = False) -> str:
+    print("starting factorio server...")
+    threadreg.append(proc)
+    line: str = ""
+    lastlines = ["", "", "", "", "", "", "", ""]
```
Suggestion: you might like `collections.deque`

suisseWalter committed:
converted everything to use Path objects
the benchmark_folder and migrate_folder can still take strings.
as these are potentially called from other files.
Also further code cleanup according to discussions in the PR.

rpdelaney commented on benchmarker.py:
```diff
@@ -139,17 +145,17 @@ def migrate_map(factorio_bin: str, map: str, inplace: bool, custom_script: str |
 
 
-def sync_mods(map: str, disable_all: bool = False) -> None:
+def sync_mods(map: PurePath, disable_all: bool = False) -> None:
```
Yay!

rpdelaney commented on benchmarker.py:
```diff
@@ -139,17 +145,17 @@ def migrate_map(factorio_bin: str, map: str, inplace: bool, custom_script: str |
         )
     else:
-        set_mod_command = os.path.join("fmm", fmm_name) + " --game-dir factorio disable"
+        set_mod_command = str(PurePath("fmm", fmm_name)) + " --game-dir factorio disable"
```
```suggestion
        set_mod_command = f"{PurePath('fmm', fmm_name)} --game-dir factorio disable"
```

rpdelaney commented on benchmarker.py:
```diff
@@ -288,25 +293,31 @@ def benchmark_folder(
+    files = [Path(file) for file in filenames] if filenames else [*Path("saves").glob(map_regex)]
+
+    factorio_path = PurePath(
+        factorio_bin if factorio_bin else PurePath("factorio", "bin", "x64", "factorio")
```
```suggestion
        factorio_bin or PurePath("factorio", "bin", "x64", "factorio")
```

rpdelaney commented on benchmarker.py:
```diff
@@ -681,7 +682,7 @@ def init_parser() -> argparse.ArgumentParser:
 
     if args.disable_mods:
-        sync_mods(map="", disable_all=True)
+        sync_mods(map=PurePath(""), disable_all=True)
```
```suggestion
        sync_mods(map=PurePath(), disable_all=True)
```
smallest nit ever

rpdelaney reviewed (approved):
Good for me :)

suisseWalter committed:
Update benchmarker.py

Co-authored-by: Ryan Delaney <1139517+rpdelaney@users.noreply.github.com>

suisseWalter committed:
Update benchmarker.py

Co-authored-by: Ryan Delaney <1139517+rpdelaney@users.noreply.github.com>
